### PR TITLE
Modify PSA related error codes and types

### PR DESCRIPTION
**Pull request by davidsaada**
PSA error codes and storage related types were changed in the PSA spec. So this PR:
- Changes the values of PSA error codes according to PSA spec, eliminating ITS specific error codes.
- Changes ITS related types to the more generic ones (`psa_storage_*`), as defined in the PSA spec.

### CODE REVIEW

File: include/psa/crypto_values.h [hunk 1]
```diff
@@ -40,25 +40,18 @@
...
 #define PSA_SUCCESS ((psa_status_t)0)
-#endif /* !defined(PSA_SUCCESS) */
 
 /** An error occurred that does not correspond to any defined
  * failure cause.
  *
  * Implementations may use this error code if none of the other standard
  * error codes are applicable. */
-#define PSA_ERROR_UNKNOWN_ERROR         ((psa_status_t)1)
+#define PSA_ERROR_GENERIC_ERROR         ((psa_status_t)-132)
```
Comment (Patater) [inline include/psa/crypto_values.h]: The allocated_errors.h header file says this error is defined in error.h. Should we move it to a new error.h file? The header seems to expect certain errors coming from headers with a certain name.

```C
#define PSA_ERROR_GENERIC_ERROR ((psa_status_t)-132) /* Defined in <psa/error.h> */
/* ... */
#define PSA_ERROR_INVALID_PADDING ((psa_status_t)-150) /* Defined in <psa/crypto.h> */
```

CC @gilles-peskine-arm Are you aware of the expectations on PSA implementations here?

File: library/psa_crypto_storage_its.c [hunk 2]
```diff
@@ -35,71 +36,37 @@
...
     return( status );
 }
 
 int psa_is_key_present_in_storage( const psa_key_id_t key )
 {
-    psa_its_status_t ret;
-    psa_its_uid_t data_identifier = psa_its_identifier_of_slot( key );
-    struct psa_its_info_t data_identifier_info;
+    psa_status_t ret;
+    psa_storage_uid_t data_identifier = psa_its_identifier_of_slot( key );
```
Comment (Patater) [inline library/psa_crypto_storage_its.c]: This commit appears to be mixing ITS type updates with error status updates. Could you please make the identifier update changes in a separate commit?

File: include/psa/crypto_values.h (hunk 1 above)
Comment (gilles-peskine-arm) [reply include/psa/crypto_values.h]: We're standardizing `psa_status_t` and a common set of status codes across PSA services, to be defined in a header file `psa/error.h` that all services provide and which Arm will manage centrally (so if you end up with several copies of `psa/error.h`, there's always a total order between versions and you can pick the most recent one). The publishing strategy is that `error.h` will be distributed by every service that uses it. See internal discussions at https://github.com/ARMmbed/PSA-IPC-doc/issues/133 and https://github.com/ARMmbed/PSA-IPC-doc/pull/177

Comment (Patater) [reply include/psa/crypto_values.h]: I'm wondering if this commit should then grab and add that error.h file instead, leaving only crypto.h specified errors in crypto.h.

The the second commit in this PR can update the ITS types.

**Comment by Patater** [comment]
This is failing the psa_constant_names tests and either this PR attempts to introduce a bug, or we need to adapt the psa_constant_names test to accommodate this PR's changes.

File: include/psa/crypto_values.h [hunk 3]
```diff
@@ -40,25 +40,18 @@
...
-#if !defined(PSA_SUCCESS)
-/* If PSA_SUCCESS is defined, assume that PSA crypto is being used
- * together with PSA IPC, which also defines the identifier
- * PSA_SUCCESS. We must not define PSA_SUCCESS ourselves in that case;
- * the other error code names don't clash. This is a temporary hack
- * until we unify error reporting in PSA IPC and PSA crypto.
- *
- * Note that psa_defs.h must be included before this header!
+/* PSA error codes */
+/* List should comply with the ones defined in https://github.com/ARMmbed/PSA-IPC-doc/blob/master/allocated_errors.h
```
[the remark this answers is not in the document]
Comment (davidsaada) [reply include/psa/crypto_values.h]: > 
> 
>     * Please break this up into meaningful commits as requested by Jaeden.
> 
>     * This breaks the ITS build (`psa_status_t` is defined twice).
> 
>     * This reveals a pre-existing defect in `psa_constant_names` and `test_psa_constant_names.py`: they don't fully support negative constants. This is signalled by CI, so we need to fix it. Fix available in #56

Will do (will also remove the private link). However, it's still not clear yet whether I should move all error codes to a separate error.h file or keep them in crypto_values.h.

**Comment by davidsaada** [comment]
> 
> 
>     * Please break this up into meaningful commits as requested by Jaeden.
> 
>     * This breaks the ITS build (`psa_status_t` is defined twice).
> 
>     * This reveals a pre-existing defect in `psa_constant_names` and `test_psa_constant_names.py`: they don't fully support negative constants. This is signalled by CI, so we need to fix it. Fix available in #56

Will do (will also remove the private link). It's still not clear whether I should move all error codes to a new `error.h` file or keep them under `crypto_values.h`

File: library/psa_crypto_storage_its.c [hunk 4]
```diff
@@ -27,6 +27,7 @@
 
 #if defined(MBEDTLS_PSA_CRYPTO_STORAGE_ITS_C)
 
+#include "psa/error.h"
```
[the remark this answers is not in the document]
Comment (davidsaada) [reply library/psa_crypto_storage_its.c]: Not sure why this would fail the build. Type definitions in both files are the same, meaning that the compiler should not fail this duplication. I build it this way in mbed-os without issues.
Is this an empirical or a theoretical problem? I ask because I only added `psa/error.h` to my current mbed-os PR, so if this is empirical, than where do you take `psa/error.h` from?

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/psa_crypto_storage_its.c]: Theoretical in this instance, but empirical from a previous version. We had to add guards around the definition of `psa_status_t` so that it would only be defined in one place. You can have multiple `#define` for the same macro as long as they expand to the same token sequence, but some compilers warn about it. You can't have multiple `typedef` statements defining the same type name, even if the definitions are identical.

In your build, what exactly does `psa/error.h` contain? Who's defining `psa_status_t` and how do you avoid defining it twice?

**Comment by davidsaada** [comment]
@Patater @gilles-peskine-arm  Commits are now split to two: one for error codes, one for the change of ITS types to the generic types.
Error codes are still in `crypto_values.h`, as it seems you're not sure whether you want to keep them there or move them into a new error.h file. It can be done later on I guess.
All comments addressed except for the `psa_status_t` duplication.

File: library/psa_crypto_storage_its.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (davidsaada) [reply library/psa_crypto_storage_its.c]: `psa/error.h` is part of the PSA include directory. I'm adding it now to [the parallel PR in mbed-os](https://github.com/ARMmbed/mbed-os/pull/9708). Please note that `psa_status_t` is not a define but a typedef. As opposed to defines, compilers will not warn about duplicate typedefs as long as the referenced type is identical. In addition, as it's a typedef and not a define, one can't ask whether it's defined using `#ifdef`.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/psa_crypto_storage_its.c]: Duplicate definitions of a typedef _are_ forbidden in C89 and C99. From N1256 (C99 TC3), §6.7 “Declarations”:

> 3\. If an identifier has no linkage, there shall be no more than one declaration of the identifier
(in a declarator or type specifier) with the same scope and in the same name space, except
for tags as specified in 6.7.2.3.

Clang does warn about it if you ask it to use C89 or C99.
```
$ cat a.c
typedef int foo;
typedef int foo;
$ clang -std=c99 -Wall -Wextra -Werror -O -c a.c
a.c:2:13: error: redefinition of typedef 'foo' is a C11 feature
      [-Werror,-Wtypedef-redefinition]
typedef int foo;
            ^
a.c:1:13: note: previous definition is here
typedef int foo;
            ^
1 error generated.
```
Neither Clang nor GCC warn about it in non-picky mode, but I remember we've hard this error in at least one of the CC+CFLAGS combination we test in our CI, and we do aim to stick to standard C in Mbed TLS and Mbed Crypto.

[the remark this answers is not in the document]
Comment (davidsaada) [reply library/psa_crypto_storage_its.c]: OK, understood (didn't go that far).
How do you suggest to resolve this? It's a bit of a pickle, as there's no `#ifdef` for typedefs. I could ask whether `PSA_SUCCESS` is defined, assuming that `PSA_SUCCESS` and `psa_status_t` are both defined in `psa/error.h`, but that just seems pretty awkward to me.

[the remark this answers is not in the document]
Comment (gilles-peskine-arm) [reply library/psa_crypto_storage_its.c]: That's the hack we had for a long time, and getting rid of that hack was my motivation for requesting a unified `psa/error.h` (https://github.com/ARMmbed/PSA-IPC-doc/issues/133) as the only place where `psa_status_t` is defined. Crypto is planning to switch from defining `psa_status_t` in `crypto_types.h` to including `error.h` as soon as version 1 of `error.h` has stabilized.

[the remark this answers is not in the document]
Comment (davidsaada) [reply library/psa_crypto_storage_its.c]: OK, will add this hack.

**Comment by davidsaada** [comment]
Added the handling of `psa_status_t` duplication. All comments are addressed now.

**Comment by davidsaada** [comment]
> 
> 
> Code LGTM
> 
> I think the commit message for "Change ITS specific types to the more generic ones (defined in PSA spec)" needs some work, though.
> 
> How about the following, but after you fill in xxx with the version of the ITS spec this PR updates to?
> 
> ```
> psa: Update types to version xxx of PSA ITS
> 
> This replaces psa_its_status_t with the more generic psa_status_t,
> psa_its_info_t with the new psa_storage_info_t, and psa_its_uid_t with
> the new psa_storage_uid_t. This is necessary in order to integrate with
> the newer implementation of PSA ITS landing in Mbed OS soon.
> ```

Cool. Commit message modified in the spirit of your recommendation. Didn't see any change to the PSA spec version though, so left this one out.

**Comment by Patater** [comment]
Sorry for the nitpicking, but it looks like you forgot to mention
```
- psa_storage_uid_t replaces psa_its_uid_t
```

Also, please don't end commit messages with a `:` or `.`, as per our [git commit style guide](https://os.mbed.com/docs/mbed-os/v5.10/reference/workflow.html#guidelines-for-github-pull-requests)

Thanks

**Comment by davidsaada** [comment]
> 
> 
> Sorry for the nitpicking, but it looks like you forgot to mention
> 
> ```
> - psa_storage_uid_t replaces psa_its_uid_t
> ```
> 
> Also, please don't end commit messages with a `:` or `.`, as per our [git commit style guide](https://os.mbed.com/docs/mbed-os/v5.10/reference/workflow.html#guidelines-for-github-pull-requests)
> 
> Thanks

This ain't nitpicking, it was a mistake I made there. Fixed now.

**Comment by davidsaada** [comment]
@Patater Thanks for approving. Can you also approve [PR #9708 in the mbed-os repo](https://github.com/ARMmbed/mbed-os/pull/9708)? It's the corresponding PR in mbed-os, including the same mbed-crypto changes.

**Comment by davidsaada** [comment]
Thanks @gilles-peskine-arm. Wasn't sure about these error codes myself. Replaced them according to your suggestion.

File: tests/suites/test_suite_psa_crypto.function [hunk 5]
```diff
@@ -422,7 +422,7 @@ static psa_status_t key_agreement_with_self( psa_crypto_generator_t *generator,
     /* Return UNKNOWN_ERROR if something other than the final call to
```
Comment (gilles-peskine-arm) [inline tests/suites/test_suite_psa_crypto.function]: Comment needs updating

File: include/psa/crypto_values.h [hunk 6]
```diff
@@ -92,23 +84,19 @@
...
 
-/** A slot is occupied, but must be empty to carry out the
- * requested action.
+/** Asking for an item that already exists
  *
- * If a handle is invalid, it does not designate an occupied slot.
- * The error for an invalid handle is #PSA_ERROR_INVALID_HANDLE.
- */
-#define PSA_ERROR_OCCUPIED_SLOT         ((psa_status_t)5)
+ * Implementations should return this error, when attempting
```
Comment (gilles-peskine-arm) [inline include/psa/crypto_values.h]: The descriptions of `PSA_ERROR_ALREADY_EXISTS`, `PSA_ERROR_DOES_NOT_EXIST` and `PSA_ERROR_INSUFFICIENT_DATA` need work (grammaticality, specific meaning for crypto) but this is not a blocker, I can fix it later.

**Comment by davidsaada** [comment]
@gilles-peskine-arm Pushed a fix according to your comments. Left the descriptions for the new error codes out (thought it would be better if an mbed crypto specialist would do that).

**Comment by davidsaada** [comment]
> 
> 
> Looks fine.
> 
> Approved provided that this has been tested (including running `test_suite_psa_crypto_entropy`) on Mbed OS in configurations with ITS with and without entropy injection. @Patater Is this tested on CI at the moment?

Thanks for the approval.
As for the tests  - I currently test it locally, but regarding CI, changes aren't currently tested yet. There's a [pending PR](https://github.com/ARMmbed/mbed-os/pull/9708) in mbed-os, which includes these changes as well. This PR awaits your approval in order to start CI. I'll copy the latest changes from here to there and then will appreciate your quick approval for it.

**Comment by Patater** [comment]
> @Patater Is this tested on CI at the moment?

No, otherwise we'd have seen a failure in CI.

**Comment by davidsaada** [comment]
@gilles-peskine-arm @Patater Sorry for nagging, but can you approve [PR #9708](https://github.com/ARMmbed/mbed-os/pull/9708) in mbed-os? The mbed-crypto part changes there are identical to the ones in this PR (they are all concentrated in the second commit). They will be discarded before that PR is merged.

**Comment by Patater** [comment]
Travis failure is DTLS 3d Proxy. Jenkins failure is FreeBSD timing test. Both are known issues.

Pre-merge test failed because this PR is not based on a version of development recent enough to have psa_constant_names that handles signed error codes (https://github.com/ARMmbed/mbed-crypto/pull/56).
